edward/util/random_variables.py

```diff
@@ -778,3 +778,66 @@ def transform(x, *args, **kwargs):
   new_x = TransformedDistribution(x, bij, *args, **kwargs)
   new_x.support = new_support
   return new_x
+
+
+def marginal(x, n):
+  """Performs a full graph sample on the provided random variable.
+
+  Given a random variable and a sample size, adds an additional sample
+  dimension to the root random variables in x's graph, and samples from
+  a new graph in terms of that sample size.
+
+  Args:
+    x : RandomVariable.
+      Random variable to perform full graph sample on.
+    n : tf.Tensor or int
+      The size of the full graph sample to take.
+
+  Returns:
+    tf.Tensor.
+    The fully sampled values from x, of shape [n] + x.shape
+
+  #### Examples
+
+  ```python
+  ed.get_session()
+  loc = Normal(0.0, 100.0)
+  y = Normal(loc, 0.0001)
+  conditional_sample = y.sample(50)
+  marginal_sample = ed.marginal(y, 50)
+
+  np.std(conditional_sample.eval())
+  0.000100221
+
+  np.std(marginal_sample.eval())
+  106.55982
+  ```
+
+  #### Notes
+
+  The current implementation only works for graphs of RVs that don't use
+  the `sample_shape` kwarg.
+  """
+  ancestors = get_ancestors(x)
+  if any([rv.sample_shape != () for rv in ancestors]) or x.sample_shape != ():
+    raise NotImplementedError("`marginal` doesn't support graphs of RVs "
+                              "with non scalar sample_shape args.")
+  elif ancestors == []:
+    old_roots = [x]
+  else:
+    old_roots = [rv for rv in ancestors if get_ancestors(rv) == []]
+
+  new_roots = []
+  for rv in old_roots:
+    new_rv = copy(rv)
+    new_rv._sample_shape = tf.TensorShape(n).concatenate(new_rv._sample_shape)
```

Comment:
`tf.TensorShape()` fails if `n` is a `tf.Tensor`

Reply:
This also came up when I looked into #774, and I think would need to be solved at the same time. Sample shape needs a TensorShape, and there's no nice way to turn tensor `n` into one (I don't think). 

So I think either `sample_shape` needs to be interpreted as 'whatever gets passed to sample' and therefore is stored as a tensor, not a tensorshape. This would solve #774, and I don't think would break much. Another alternative is having `sample_shape` and `sample_shape_tensor` attributes built from the actual tensor representation of the RV.

Let me know if you'd prefer this is implemented in the same PR, I'll push the other changes.